File: src/glimmung/public_ids.py
"""Public, human-readable handles for operator-facing contracts.

Storage rows still keep their Cosmos `id` values internally. Helpers in this
module are for API/MCP/frontend surfaces where exposing backing IDs would make
the storage key part of the product contract.
"""

from __future__ import annotations


def issue_ref(project: str, number: int | None) -> str:
    if number is None:
        return project
    return f"{project}#{number}"


def run_ref(project: str, issue_number: int | None, run_display: str | int | None) -> str:
    run_part = str(run_display).strip() if run_display is not None else ""
    if issue_number is None:
        return f"{project}/runs/{run_part or 'unknown'}"
    return f"{project}#{issue_number}/runs/{run_part or 'unknown'}"


def report_ref(repo: str, number: int | None) -> str:
    if number is None:
        return repo
    return f"{repo}#{number}"


def lease_ref(project: str, *, slot_name: str | None = None, lease_number: int | None = None) -> str:
    if slot_name:
        return slot_name
    if lease_number is not None:
        return f"{project}/leases/{lease_number}"
    return f"{project}/lease"

File: internal/domain/publicids/publicids.go
package publicids

import (
	"strconv"
	"strings"
)

// IssueRef returns the operator-facing issue reference for a project issue.
func IssueRef(project string, number *int) string {
	if number == nil {
		return project
	}
	return project + "#" + strconv.Itoa(*number)
}

// RunRef returns the operator-facing run reference.
func RunRef(project string, issueNumber *int, runDisplay string) string {
	runPart := strings.TrimSpace(runDisplay)
	if runPart == "" {
		runPart = "unknown"
	}
	if issueNumber == nil {
		return project + "/runs/" + runPart
	}
	return project + "#" + strconv.Itoa(*issueNumber) + "/runs/" + runPart
}

// TouchpointRef returns the operator-facing touchpoint or pull request reference.
func TouchpointRef(repo string, number *int) string {
	if number == nil {
		return repo
	}
	return repo + "#" + strconv.Itoa(*number)
}

// LeaseRef returns the operator-facing lease reference.
func LeaseRef(project, slotName string, leaseNumber *int) string {
	if slotName != "" {
		return slotName
	}
	if leaseNumber != nil {
		return project + "/leases/" + strconv.Itoa(*leaseNumber)
	}
	return project + "/lease"
}
